python/packages/ai/teams/app.py

```diff
@@ -202,13 +205,97 @@ def __selector__(context: TurnContext):
 
             return False
 
-        def __call__(func: Callable[[TurnContext, StateT], Awaitable[bool]]):
+        def __call__(func: RouteHandler[StateT]):
+            self._routes.append(Route[StateT](__selector__, func))
+            return func
+
+        return __call__
+
+    def anonymous_query_link(self, command_id: Union[str, Pattern[str]]):
+        """
+        Registers a handler for a command that performs anonymous link unfurling.
+
+        ```python
+        # Use this method as a decorator
+        @app.anonymous_query_link("test")
+        async def on_anonymous_query_link(context: TurnContext, state: TurnState):
+            print(context.activity.value.url)
+            return True
+
+        # Pass a function to this method
+        app.anonymous_query_link("test")(on_anonymous_query_link)
+        ```
+
+        #### Args:
+        - `command_id`: a string or regex pattern that matches against the activities
+        `command_id`
+        """
+
+        def __selector__(context: TurnContext):
+            if (
+                context.activity.type != ActivityTypes.invoke
+                or context.activity.name != "composeExtension/anonymousQueryLink"
+                or not context.activity.value
+            ):
+                return False
+
+            return self._activity_with_command_id(context.activity, command_id)
+
+        def __call__(func: RouteHandler[StateT]):
+            self._routes.append(Route[StateT](__selector__, func))
+            return func
+
+        return __call__
+
+    def message_preview(self, command_id: Union[str, Pattern[str]], action: MessagePreviewAction):
+        """
+        Registers a handler to process an action of a message that's being
+        previewed by the user prior to sending.
+
+        ```python
+        # Use this method as a decorator
+        @app.message_preview("test", "edit")
+        async def on_message_preview_edit(context: TurnContext, state: TurnState):
+            return True
+
+        # Pass a function to this method
+        app.message_preview("test", "edit")(on_message_preview_edit)
+        ```
+
+        #### Args:
+        - `command_id`: a string or regex pattern that matches against the activities
+        `command_id`
+        - `action`: the action to subscribe to
+        """
+
+        def __selector__(context: TurnContext):
+            if (
+                context.activity.type != ActivityTypes.invoke
+                or context.activity.name != "composeExtension/submitAction"
+                or not context.activity.value
+                or not self._activity_with_command_id(context.activity, command_id)
+            ):
+                return False
+
+            message_preview_action = None
+
+            if isinstance(context.activity.value, object):
```

Comment:
minor: the coding style is a bit different than `anonymous_query_link`. `anonymous_query_link` wraps the selector logic into a new function. You can see whether to unify the style.

Reply:
I do not see any difference, they are both using `__selector__` and wrapping in `__call__` the same exact way.